### PR TITLE
The Complete Guide to Deploying JavaScript Applications - Part 1:  Database, and API Deployment

### PR DESCRIPTION
**Pull request by unicodeveloper**

### CODE REVIEW

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+
+> Heroku runs your Node.js app in a dyno, a smart container which provides a modern stack with your choice of web server and runtime.
+
+Clone the backend from [GitHub](https://github.com/auth0-blog/jsbackend). 
+
+Heroku looks for a [Procfile](https://devcenter.heroku.com/articles/procfile). If no `Procfile` exists for the Node.js app, Heroku wull attempt to start a default web process via the `start script` in your `package.json`. In our app's backend, we have a `start script`.
+
+![JavaScript Deploy - Start Script](https://cdn.auth0.com/blog/jsdeploy/startscript.png)
+
+>Note: A Procfile is a text file in the root directory of your application that defines process types and explicitly declares what command should be executed to start your app on heroku.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: heroku -> Heroku

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+
+Make sure you follow these steps below:
+
+- Create an account with [MongoDB Atlas](https://www.mongodb.com/cloud/atlas)
+- A dialog box will be presented to you with a lot of form fields.
+- Choose a name for your cluster, a cloud provider, and region like so:
+  
+  ![JavaScript Deploy - Cluster details](https://cdn.auth0.com/blog/jsdeploy/atlasclusterdetails.png)
+
+- Scroll down a bit to **Instance size**. Here you can choose a plan for your app. I chose the free version for the purpose of this tutorial. The free version is also great for prototyping, POCs and demos.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: POCs and demos -> POCs, and demos

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+
+- Flexible, hierarchical data structures.
+- Stores data in documents, organized into collections.
+- Queries for retrieving individual, specific documents or to retrieve all the documents in a collection that match your query parameters.
+- Realtime updates via data synchronization.
+- Offline usage.
+
+Follow the steps below to create a database.
+
+- Signup for an [account on Firebase](https://console.firebase.google.com/).
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: Signup -> Sign up

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 4]
```diff
@@ -0,0 +1,717 @@
...
+It offers:
+
+- Up to 10TB of storage capacity, 40,000 IOPS, and 416GB of RAM per instance.
+- Automatic data encryption.
+- Google Cloud SQL is SSAE 16, ISO 27001, PCI DSS v3.0, and HIPAA compliant.
+- High performance and more.
+
+You can follow the steps below to get started with provisioning a database.
+
+- Head over to [Google Cloud SQL](https://console.cloud.google.com/sql/). If you don't have a google acount, sign up for one.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: acount -> account

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 5]
```diff
@@ -0,0 +1,717 @@
...
+Amazon Relational Database Service (RDS) is a cloud platform that makes it easy to set up a relational database. It provides six popular database engines:
+
+- [Aurora](https://aws.amazon.com/rds/aurora/)
+- [PostgreSQL](https://aws.amazon.com/rds/postgresql/)
+- [MySQL](https://aws.amazon.com/rds/mysql/)
+- [MariaDB](https://aws.amazon.com/rds/mariadb/)
+- [Oracle](https://aws.amazon.com/rds/oracle/)
+- [Microsoft SQL Server](https://aws.amazon.com/rds/sqlserver/)
+
+It is higly optimized for memory, I/O and performance.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: higly -> highly

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 6]
```diff
@@ -0,0 +1,717 @@
...
+
+## Code Deployment
+
+We have briefly discussed the several ways and services you can deploy your database. Now, let's talk about the code. Here, we'll deploy the backend. There are several awesome cloud platforms that do a very awesome job at backend deployments. 
+
+### Heroku
+
+Heroku is a cloud platform that helps you deploy and host your applications the modern way. It does all the heavy-lifting for you. Let's quickly take a look at how to deploy and maintain a Node.js application on heroku.
+
+Heroku runs a first class experience for Node.js so every developer can run successful production apps.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: first class experience -> first-class experience

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 7]
```diff
@@ -0,0 +1,717 @@
...
+
+If you don't have an account, go ahead and create one on [heroku.com](https://www.heroku.com/). Go ahead and install the [heroku cli](https://devcenter.heroku.com/articles/heroku-cli).
+
+> Heroku runs your Node.js app in a dyno, a smart container which provides a modern stack with your choice of web server and runtime.
+
+Clone the backend from [GitHub](https://github.com/auth0-blog/jsbackend). 
+
+Heroku looks for a [Procfile](https://devcenter.heroku.com/articles/procfile). If no `Procfile` exists for the Node.js app, Heroku wull attempt to start a default web process via the `start script` in your `package.json`. In our app's backend, we have a `start script`.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: wull -> will

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 8]
```diff
@@ -0,0 +1,717 @@
...
+![Build Succeeded](https://cdn.auth0.com/blog/jsdeploy/heroku-build.png)
+_Build Succeeded_
+
+![New version of app](hhttps://cdn.auth0.com/blog/jsdeploy/index-app-route.png)
+_New version of the app showing the index route_
+
+
+### Caching & Cron Jobs
+
+Heroku provides an array of addons for caching, from `memcache` to `fastly` to `ironcache` and others. You can check out how to use [memcache with PHP on heroku](https://devcenter.heroku.com/articles/memcachedcloud#using-memcached-from-php).
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: Link to Fastly or SEO? :)

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 9]
```diff
@@ -0,0 +1,717 @@
...
+
+Finally, you can use the [Heroku Scheduler](https://devcenter.heroku.com/articles/scheduler) for running jobs on your app at scheduled time intervals.
+
+## Google Cloud
+
+The [Google Cloud platform](https://cloud.google.com) is a trusted cloud platform that lots of companies and startups utilize in deploying and hosting their apps. The apps deployed on `Google Cloud` run on the same infrastructure that powers all of Google's products.
+
+Google Cloud offers different options for hosting Node.js apps. The platform offers [App Engine (Full managed)](https://cloud.google.com/appengine/), [Compute Engine (Scalabe VMs)](https://cloud.google.com/compute/) and [Container Engine (Kubernetes Clusters)](https://cloud.google.com/container-engine/).
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: Scalabe VMs -> Scalable VMs

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 10]
```diff
@@ -0,0 +1,717 @@
...
+
+In our case, we'll use Google App Engine to deploy our Node.js backend. It abstracts the infrastucture away. Let's dive right in!
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: infrastucture -> infrastructure

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 11]
```diff
@@ -0,0 +1,717 @@
...
+
+When using Google App Engine, you can choose the **Standard** or **Flexible** environment. The latter, like the name implies allows you to install a lot of extensions and allows you to run deployment scripts using `package.json`. We'll use the flexible environment. So, go ahead and [create a new project](https://console.cloud.google.com/projectselector/appengine/create). Click on **Create**, give the project a name, select the region you'll like to serve your app and enable billing.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: The latter, like the name implies allows -> The latter, like the name implies, allows

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 12]
```diff
@@ -0,0 +1,717 @@
...
+gcloud config set project <projectID>
+```
+
+Now go ahead and run the following command to deploy the app:
+
+```bash
+gcloud app deploy
+```
+
+Our backend is finally [live.](https://meetupservice-190908.appspot.com).
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: live. -> live

. is after the link already

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 13]
```diff
@@ -0,0 +1,717 @@
...
+- Call services inside or outside of Azure.
+- Run jobs on any schedule.
+- Use Azure Storage queues for long-running or offline jobs.
+- Invoke Azure Service Bus queues.
+
+Check out [how to create and manage jobs using the Scheduler](https://docs.microsoft.com/en-us/azure/scheduler/scheduler-get-started-portal).
+
+## Amazon Web Services
+
+More companies use AWS(Amazon Web Services) for storing all sorts of data ranging from images, mp3 files to videos than any other cloud platform. In fact, a lot of organizations like Uber, Spotify, Salesforce use Amazon Web Services wholy for hosting, deployment and infrastructure. AWS has a ton of developer products.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: wholy -> wholly

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 14]
```diff
@@ -0,0 +1,717 @@
...
+
+The service we'll use for deploying our backend Node.js app is **Amazon Elastic Beanstalk**. Let's get started.
+
+- Sign up for an [AWS account](https://aws.amazon.com/account/) if you don't have one.
+- Head over to [Elastic Beanstalk console](https://us-west-2.console.aws.amazon.com/elasticbeanstalk).
+- Create a new app.
+    ![Create new app](https://cdn.auth0.com/blog/jsdeploy/awscreatenewapp.png)
+- Click on `create web server`.
+    ![Create web server](https://cdn.auth0.com/blog/jsdeploy/awscreatewebserver.png)
+- Create the webserver environment.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: webserver -> web server

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 15]
```diff
@@ -0,0 +1,717 @@
...
+- Upload your code. Elastic Beanstalk requires that you upload a zip file of your codebase. You can manually zip it up. But I prefer to do that from my terminal like so:
+
+    ```bash
+     zip ../meetupservice.zip -r * .[^.]*
+    ```
+- Now, upload it to AWS like so:
+    ![Upload code to AWS](https://cdn.auth0.com/blog/jsdeploy/awsuploadcode.png)
+- Check availability for the app URL. Mine looks like this:
+    ![URL](https://cdn.auth0.com/blog/jsdeploy/awsavailableurl.png)
+- The next page allows us to configure Database Instance. Now our app already uses a remote mongoDB instance, so we can skip it.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: mongoDB -> MongoDB

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 16]
```diff
@@ -0,0 +1,717 @@
...
+### Caching and Cron Jobs
+
+For caching, Amazon Web Services offers [ElastiCache](https://aws.amazon.com/elasticache). It is a web service that makes it easy to deploy, operate, and scale an in-memory data store or cache in the cloud. Amazon ElastiCache supports two open-source in-memory engines:
+
+- [Redis](https://aws.amazon.com/elasticache/redis).
+- Memcached.
+
+Amazon ElastiCache automatically detects and replaces failed nodes, reducing the overhead associated with self-managed infrastructures and provides a resilient system that mitigates the risk of overloaded databases, which slow website and application load times. Through integration with [Amazon CloudWatch](https://aws.amazon.com/cloudwatch), Amazon ElastiCache provides enhanced visibility into key performance metrics associated with your Redis or Memcached nodes.
+
+Companies like [AirBnb](https://www.airbnb.com), [Healthguru](http://www.healthguru.com), [PlaceIQ](http://www.placeiq.com) and [Tokyo Data Network](http://www.tdn.co.jp) use ElastiCache for caching at multiple layers spanning HTML fragments, results of expensive DB queries, ephemeral session data and search results.
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: Airbnb

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 17]
```diff
@@ -0,0 +1,717 @@
...
+## Application To Deploy 
+
+In this tutorial, we'll be deploying this sample application, **Meetups**. The Architecture is not complex. It's composed of a back-end and a front-end.
+
+The back-end comprises of the following:
+
+- **A MongoDB Database**
+- **A Node.js web framework - Express**
+
+The front-end comprises of the following:
```
[the remark this answers is not in the document]
Comment (kmaida) [reply _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: is comprised of

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 18]
```diff
@@ -0,0 +1,717 @@
...
+
+- **Vue.js** - JavaScript Front-end framework.
```
[the remark this answers is not in the document]
Comment (kmaida) [reply _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: is comprised of

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 19]
```diff
@@ -0,0 +1,717 @@
...
+
+- **Front-end:** _HTML_, _CSS_, and _JavaScript_. A couple of CSS frameworks that can be used on the front-end are:
+
+  <ul>
+    <li>**[Bootstrap](https://getbootstrap.com/):** The most popular CSS framework in the world </li>
+    <li>**[Bulma](https://bulma.io):** A modern CSS framework based on Flexbox </li>
+    <li>**[Tailwind](https://tailwindcss.com/):** A utility-first CSS framework for rapid UI development </li>
+  </ul>
+
+A couple of JavaScript front-end frameworks that can be used on the front-end are:
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: This line should be tabbed in 2 spaces for the indentation to carry through it and the `<ul>` below

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 20]
```diff
@@ -0,0 +1,717 @@
...
+- **Back-end:** JavaScript running on a Node.js server. Some frameworks that can be used on the back-end are:
+
+  <ul>
+    <li>**[Express](https://expressjs.com/):** Fast, unopinionated, minimalist web framework for Node.js. </li>
+    <li>**[Hapi](https://hapijs.com/):** Rich framework for building applications and services. </li>
+    <li>**[Koa](http://koajs.com/):** Next generation web framework for Node.js. </li>
+    <li>**[Sails](https://sailsjs.com/):** Realtime MVC framework for Node.js. </li>
+  </ul>
+
+Examples of Databases that can be used are:
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: Same for this line, tab in 2 spaces

File: _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown [hunk 21]
```diff
@@ -0,0 +1,717 @@
...
+});
+...
+```
+
+Commit and push to your master branch. Now, go to the **Activity** tab of your Heroku Dashboard and notice the build. Reload your app and you'll see the difference.
+
+![Build Succeeded](https://cdn.auth0.com/blog/jsdeploy/heroku-build.png)
+_Build Succeeded_
+
+![New version of app](hhttps://cdn.auth0.com/blog/jsdeploy/index-app-route.png)
```
Comment (lukeoliff) [inline _posts/2017-12-21-the-complete-guide-to-deploying-javascript-applications-part-1.markdown]: hhttps://cdn.auth0.com/blog/jsdeploy/index-app-route.png -> https://cdn.auth0.com/blog/jsdeploy/index-app-route.png

Getting a 404 on render :)

**Comment by lukeoliff** [comment]
I also spotted some frontend/backend uses that should also be front end/back end

**Comment by unicodeveloper** [comment]
Yeah. I already fixed that just now. Will push the changes @lukeoliff

**Pull request merged**